### PR TITLE
Feature/addition rest endpoints

### PR DESCRIPTION
**Pull request by lliss**
Connects #7 

Builds off of https://github.com/ProboCI/probo-asset-receiver/pull/6 https://github.com/ProboCI/probo-asset-receiver/pull/8 and https://github.com/ProboCI/probo-asset-receiver/pull/9

Adds additional API endpoints needed to allow users to review the size of their assets and remove them if they so desire.

NOTE: Only the last 5 commits are part of this PR.

### To Test:
 - You will first need to run the migrations (see testing instructions https://github.com/ProboCI/probo-asset-receiver/pull/6)
 - Upload new files
 - View assets in your bucket at `/buckets/:bucketName/assets`
 - Ensure the sizes and other attributes are appropriate for the file you uploaded.
 - Delete the file via the new endpoint `/buckets/:bucketName/assets/:assetName`
 - Ensure that the db keys and the files on disk are deleted.

### CODE REVIEW

File: lib/Server.js [hunk 1]
```diff
@@ -1,28 +1,30 @@
...
+var through2 = require('through2');
 
 var authlib = require('./auth');
+var plugins = require('./plugins');
+var ByteCounter = require('./ByteCounter');
 
 var Server = function(options) {
   this.options = options || {
     encryptionCipher: 'aes-256-cbc',
     encryptionPassword: 'SECRET',
```
Comment (tortillaj) [inline lib/Server.js]: Let's make this `encryptionPassword` a config option.

Comment (tortillaj) [inline lib/Server.js]: Oh I see now, never mind this comment ^

File: lib/Server.js [hunk 2]
```diff
@@ -77,6 +75,9 @@ Server.prototype.setup = function() {
   app.delete('/buckets/:bucket/token/:token', auth, this.routes.deleteBucketToken);
   app.get('/buckets/:bucket/token', auth, this.routes.listBucketTokens);
   app.get('/asset/:bucket/:assetName', auth, this.routes.serveFileAsset);
+  app.get('/asset-size/:bucket/:assetName', auth, this.routes.listAssetSize);
```
Comment (tortillaj) [inline lib/Server.js]: This is kind of nit-picky, but I'd like to propose `/buckets/:bucket/asset/:assetName/size` just to make the URL more predictable.

File: test/awsS3-test.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
+'use strict';
```
Comment (tortillaj) [inline test/awsS3-test.js]: this file needs linting

File: migrations/probo-asset-fix-and-upload [hunk 4]
```diff
@@ -0,0 +1,104 @@
+#! /usr/bin/env node
```
Comment (tortillaj) [inline migrations/probo-asset-fix-and-upload]: Can we rename the migrations to include a number indicating the order they need to be run? e.g. `0001-probo-asset-fix-and-upload`

Comment (tizzo) [inline migrations/probo-asset-fix-and-upload]: Yeah, we should do that. Also - I think we should probably title this
migration to be a bit more descriptive. I think this is
"0000-get-upload-sizes-and-reencrypt' or something?

Comment (lliss) [inline migrations/probo-asset-fix-and-upload]: Sure. No problem.

File: test/awsS3-test.js (hunk 3 above)
Comment (lliss) [inline test/awsS3-test.js]: Ok.

File: lib/Server.js (hunk 2 above)
Comment (lliss) [inline lib/Server.js]: I'm fine with that.

File: lib/Server.js [hunk 5]
```diff
@@ -69,6 +62,11 @@ Server.prototype.setup = function() {
     res.send('Probo asset receiver');
   });
 
+
+  app.get('/', function(req, res) {
```
Comment (tortillaj) [inline lib/Server.js]: I know it's just a one off route, but we should define it like the others.

File: lib/Server.js [hunk 6]
```diff
@@ -77,6 +75,9 @@ Server.prototype.setup = function() {
   app.delete('/buckets/:bucket/token/:token', auth, this.routes.deleteBucketToken);
   app.get('/buckets/:bucket/token', auth, this.routes.listBucketTokens);
   app.get('/asset/:bucket/:assetName', auth, this.routes.serveFileAsset);
+  app.get('/asset-size/:bucket/:assetName', auth, this.routes.listAssetSize);
+  app.get('/buckets/:bucket/assets', auth, this.routes.assetMetadataByBucket);
+  app.del('/buckets/:bucket/asset/:assetName', auth, this.routes.deleteAssetFromBucket);
```
Comment (tortillaj) [inline lib/Server.js]: `app.del` is deprecated in favor of `app.delete`

**Comment by tortillaj** [comment]
I'd like to provide a suggestion on the migrations too. It might be nice to include some kind of feedback to the console to inform if things are working ... and maybe count the files in the old file directory to compare against the new, reciphered file directory

**Comment by lliss** [comment]
> I'd like to provide a suggestion on the migrations too. It might be nice to include some kind of feedback to the console to inform if things are working ... and maybe count the files in the old file directory to compare against the new, reciphered file directory

That's a good idea.

**Comment by lliss** [comment]
I added another script that cleans up bad keys if they still exist after the other migrations. I also addressed all the feedback noted above. Can we give it another pass and then merge?

**Comment by tortillaj** [comment]
@tizzo this looks good to me!

**Comment by tortillaj** [comment]
oops - side note, I did not test AWS uploads!

**Pull request merged**